test/e2e/lib/helpers/broker_event_transformation_test_with_source_helper.go

```diff
@@ -0,0 +1,331 @@
+package helpers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	v1 "k8s.io/api/core/v1"
+	eventingv1alpha1 "knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	eventingtestlib "knative.dev/eventing/test/lib"
+	eventingtestresources "knative.dev/eventing/test/lib/resources"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/test/helpers"
+
+	// The following line to load the gcp plugin (only required to authenticate against GKE clusters).
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	kngcptesting "github.com/google/knative-gcp/pkg/reconciler/testing"
+	"github.com/google/knative-gcp/test/e2e/lib"
+	"github.com/google/knative-gcp/test/e2e/lib/resources"
+)
+
+func BrokerEventTransformationTestHelper(t *testing.T, client *lib.Client, brokerURL url.URL, brokerName string) {
+	senderName := helpers.AppendRandomString("sender")
+	targetName := helpers.AppendRandomString("target")
+
+	// Create a target Job to receive the events.
+	makeTargetJobOrDie(client, targetName)
+
+	createTriggersAndKService(client, brokerName, targetName)
+
+	// Just to make sure all resources are ready.
+	time.Sleep(5 * time.Second)
+
+	// Create a sender Job to sender the event.
+	senderJob := resources.SenderJob(senderName, []v1.EnvVar{{
+		Name:  "BROKER_URL",
+		Value: brokerURL.String(),
+	}})
+	client.CreateJobOrFail(senderJob)
+
+	// Check if dummy CloudEvent is sent out.
+	if done := jobDone(client, senderName, t); !done {
+		t.Error("dummy event wasn't sent to broker")
+		t.Failed()
```

Comment:
Why call `t.Failed()`? I think it only returns true of false??

Reply:
I refactored the code. This is what we used to use https://github.com/google/knative-gcp/blob/bb7cfe32ebeb07932f0143a1a7a4092dd218195a/test/e2e/test_broker_pubsub.go#L88. I thought when the tests fail, we will use t.Failed to mark `Fail` for this test case.

Reply:
t.Failed() does not mark the test to fail, it only return the status.

t.Error already marks the test fail.

Reply:
`t.Error()` use `c.Fail()` but `t.Failed()` call `c.Failed()`. I think previous we use t.Failed to mark failed reports and stops the execution.
// Fail marks the function as having failed but continues execution.
func (c *common) Fail() {
	if c.parent != nil {
		c.parent.Fail()
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	// c.done needs to be locked to synchronize checks to c.done in parent tests.
	if c.done {
		panic("Fail in goroutine after " + c.name + " has completed")
	}
	c.failed = true
}

// Failed reports whether the function has failed.
func (c *common) Failed() bool {
	c.mu.RLock()
	failed := c.failed
	c.mu.RUnlock()
	return failed || c.raceErrors+race.Errors() > 0
}